external/llvm-project/mlir/lib/ExecutionEngine/RocmDeviceName.cpp

```diff
@@ -0,0 +1,76 @@
+//===- RocmDeviceName.cpp - MLIR to C++ option parsing ---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements isa name string splitter
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/ExecutionEngine/RocmDeviceName.h"
+#include "llvm/Support/Error.h"
+
+using namespace mlir;
+
+static constexpr const char kGcnDefaultTriple[] = "amdgcn-amd-amdhsa";
+
+// This function converts the arch name to target features:
+// sramecc+:xnack- to +sramecc,-xnack
+static LogicalResult parseTargetFeatures(ArrayRef<StringRef> tokens,
+                                         SmallString<32> &gcnArchFeatures) {
+  // Convert each feature and append to feature result
+  bool first = true;
+  SmallVector<SmallString<8>, 2> featureTokens;
+  for (auto &token : tokens) {
+    if (!first)
+      gcnArchFeatures += ",";
+    first = false;
+    char modifier = token.back();
+    if (modifier != '+' && modifier != '-') {
+      llvm::errs() << "Malformed token: must end with +/-.\n";
+      return failure();
+    }
+    gcnArchFeatures += modifier;
+    gcnArchFeatures += token.substr(0, token.size() - 1);
+  }
+  return success();
+}
+
+RocmDeviceName::RocmDeviceName(StringRef devName)
+    : status(failure()), triple(kGcnDefaultTriple) {
```

Comment:
Okay, I have to say I dislike the idiom here of trying to make constructor do all the work. From my background in oop, constructor is for basic setup, not full computation. Instead, build the fields on demand. 

In particular I dislike about the class carrying a side effect (the `status`) information in order to capture the failing case in construction. I doubt anyone that contributes to a client of this splitter are expected to handle the possibility of construction failures.

I tend to ask you to move this into a function like `init()` that populate those fields.

Reply:
This is just a Helper utility object but yes it can fail without the user knowing. Will update.